libs/esp32boot/esp32init.erl

```diff
@@ -0,0 +1,102 @@
+%
+% This file is part of AtomVM.
+%
+% Copyright 2023 Davide Bettio <davide@uninstall.it>
+%
+% Licensed under the Apache License, Version 2.0 (the "License");
+% you may not use this file except in compliance with the License.
+% You may obtain a copy of the License at
+%
+%    http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS,
+% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+% See the License for the specific language governing permissions and
+% limitations under the License.
+%
+% SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+%
+
+-module(esp32init).
+
+-export([start/0]).
+
+start() ->
+    console:print(<<"AtomVM init.\n">>),
+    boot().
+
+is_dev_mode_enabled(SystemStatus) ->
+    case {SystemStatus, esp:nvs_get_binary(atomvm, dev_mode)} of
+        {_, <<"always">>} ->
+            true;
+        {_, <<"never">>} ->
+            false;
+        {ok, undefined} ->
+            false;
+        {app_fail, undefined} ->
+            true
+    end.
+
+maybe_start_dev_mode(SystemStatus) ->
+    case is_dev_mode_enabled(SystemStatus) of
+        true -> esp32devmode:start_dev_mode();
+        false -> not_started
+    end.
+
+% TODO: add support for multiple apps
+% /dev/partition/by-name/app1.avm
+% /dev/partition/by-name/app2.avm
+
+boot() ->
+    BootPath = get_boot_path(),
+    case atomvm:add_avm_pack_file(BootPath, [{name, app}]) of
+        ok ->
+            StartModule = get_start_module(),
+            DevOnFail = is_dev_mode_enabled(app_fail),
```

Comment:
I believe this line is wrong.  It should be passing `ok`.  If you pass `app_fail`, then the ESP will start in dev mode, even if you have installed an application (and don't want dev mode).

Test:

```
  553  esptool.py --chip esp32 --port /dev/tty.usbserial-0001 --baud 921600 erase_flash
  554  idf.py build
  555  ./build/mkimage.sh 
  556  ./build/flashimage.sh -p /dev/tty.usbserial-0001 -b 921600
  557  idf.py monitor -p /dev/tty.usbserial-0001 
```

works as expected -- it starts the device in dev mode, because there is no application flashed.

But then flash an application:
```
  558  ./build/flash.sh -p /dev/tty.usbserial-0001 -b 921600 ../../../build/examples/erlang/hello_world.avm 
  559  idf.py monitor -p /dev/tty.usbserial-0001 
```

In this case I expect hello world to run and then the VM exits.  But in this case, dev mode is started after hello_world terminates:

```
---
AtomVM init.
I (907) sys: Loaded BEAM partition main.avm at address 0x210000 (size=1048576 bytes)
Starting application...
Hello World
Exited: ok.
Starting network...
I (1037) network_driver: AP ssid: AtomVM-ESP32
I (1037) network_driver: AP authmode: 4
I (1037) network_driver: AP ssid_hidden: 0
I (1047) network_driver: AP max_connection: 4
I (1057) wifi:wifi driver task: 3ffcb79c, prio:23, stack:6656, core=0
I (1057) system_api: Base MAC address is not set
I (1057) system_api: read default base MAC address from EFUSE
I (1067) wifi:wifi firmware version: 0f80fa0
I (1067) wifi:wifi certification version: v7.0
I (1077) wifi:config NVS flash: enabled
I (1077) wifi:config nano formating: disabled
I (1087) wifi:Init data frame dynamic rx buffer num: 32
I (1087) wifi:Init management frame dynamic rx buffer num: 32
I (1097) wifi:Init management short buffer num: 32
I (1097) wifi:Init dynamic tx buffer num: 32
I (1097) wifi:Init static rx buffer size: 1600
I (1107) wifi:Init static rx buffer num: 10
I (1107) wifi:Init dynamic rx buffer num: 32
I (1117) wifi_init: rx ba win: 6
I (1117) wifi_init: tcpip mbox: 32
I (1117) wifi_init: udp mbox: 6
I (1127) wifi_init: tcp mbox: 6
I (1127) wifi_init: tcp tx win: 5744
I (1137) wifi_init: tcp rx win: 5744
I (1137) wifi_init: tcp mss: 1440
I (1137) wifi_init: WiFi IRAM OP enabled
I (1147) wifi_init: WiFi RX IRAM OP enabled
I (1147) network_driver: WIFI mode set to 2
I (1897) network_driver: AP mode configured
I (1907) phy_init: phy_version 4670,719f9f6,Feb 18 2021,17:07:07
I (2007) wifi:mode : softAP (30:c6:f7:29:62:7d)
I (2017) wifi:Total power save buffer number: 16
I (2017) wifi:Init max length of beacon: 752/752
I (2017) wifi:Init max length of beacon: 752/752
I (2017) network_driver: WIFI started
I (2017) network_driver: WIFI_EVENT_AP_START received.
WLAN AP ready. Waiting connections.
Web server listening on port 8080
ALISP console listening on port 2323
Waiting to accept shell connection...
```

Reply:
Fixed and tested.